### PR TITLE
[DEV - 2725] Added window operator mock client implementation

### PR DESCRIPTION
**Pull request by nonibansal**
Added window operator mock client implementation.
Added client tests.

Please review this https://github.com/fennel-ai/client/pull/337 first.

### CODE REVIEW

File: fennel/test_lib/executor.py [hunk 1]
```diff
@@ -458,3 +459,99 @@ def visitFirst(self, obj):
...
+            current_session_end_time = None
+            event_count = 0
+
+            # Iterate through the rows of the DataFrame
+            for index, row in df.iterrows():
+                # Check if it's the first event for this user
+                if current_session_start_time is None:
+                    current_session_start_time = row[timestamp_col]
+                    current_session_end_time = row[timestamp_col]
+                    event_count = 1
```
Comment (aditya-nambiar) [inline fennel/test_lib/executor.py:490]: I think we could actually create a Session/Window struct that has the apis, add_event and automatically increments event count, and sets start/end timestamp appropriately. We can then use this at line 506 and append it to the session list. 

This struct will also help us later when we want to store the list of events / do tumbling windows etc

Finally when creating the dataframe we can convert it to dictionary using some list comprehension syntax.

File: fennel/test_lib/mock_client.py [hunk 2]
```diff
@@ -150,19 +149,41 @@ def dataset_lookup_impl(
...
+                # Changing dtype of Struct to str for making it hashable
+                if col in right_df and right_df[col].dtype == object:
+                    cols_to_replace.append(col)
+                    right_df[f"{col}__internal"] = right_df[col].apply(
+                        lambda x: str(dict(x))
+                    )
+                    keys[f"{col}__internal"] = keys[col].apply(
+                        lambda x: str(dict(x))
+                    )
+            right_df = right_df.drop(columns=cols_to_replace)
```
Comment (aditya-nambiar) [inline fennel/test_lib/mock_client.py:166]: Is this because window is a struct and also a key ?

File: fennel/test_lib/mock_client.py [hunk 3]
```diff
@@ -188,18 +209,40 @@ def dataset_lookup_impl(
         right_df = data[cls_name]
         right_df[FENNEL_LOOKUP] = True
         right_df[FENNEL_TIMESTAMP] = right_df[timestamp_field]
+        cols_to_replace = []
         for col in keys:
             if col in right_df and keys[col].dtype != right_df[col].dtype:
                 keys[col] = keys[col].astype(right_df[col].dtype)
+            # Changing dtype of Struct to str for making it hashable
```
Comment (aditya-nambiar) [inline fennel/test_lib/mock_client.py:216]: The lookup function has now become extremely complex and over 150 lines long. Might be a good time to break it up, and maybe move it into a module where you can reuse code, the if and else branch can clearly use some DRY

**Review by aditya-nambiar — COMMENTED** [review]
Overall looks great

File: fennel/test_lib/mock_client.py (hunk 2 above)
Comment (nonibansal) [reply fennel/test_lib/mock_client.py:166]: yup :(, saw this doing in some mock implementation.

File: fennel/client_tests/test_window_operator.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    assert df_session.shape[0] == 1
+    assert df_session["user_id"].values == [1]
+    assert df_session["window"].values[0].begin == datetime(
+        2023, 1, 16, 11, 0, 25
+    )
+    assert df_session["window"].values[0].end == datetime(
+        2023, 1, 16, 11, 0, 33
+    )
+    assert df_session["window"].values[0].count == 8
```
[the remark this answers is not in the document]
Comment (nonibansal) [reply fennel/client_tests/test_window_operator.py:197]: Moved.

File: fennel/test_lib/executor.py (hunk 1 above)
Comment (nonibansal) [reply fennel/test_lib/executor.py:490]: Added the impl.

File: fennel/test_lib/mock_client.py (hunk 3 above)
Comment (nonibansal) [reply fennel/test_lib/mock_client.py]: For now I have some code to a different function. Will pick a task to modularize the code.

File: fennel/test_lib/mock_client.py [hunk 5]
```diff
@@ -72,6 +71,61 @@ def __init__(self, status_code: int, content: str):
         )
 
 
+def _dataframe_lookup(
+    dataset_name: str,
+    keys: pd.DataFrame,
+    right_df: pd.DataFrame,
+    join_columns: List[str],
+    timestamp_field: str,
+) -> pd.DataFrame:
```
Comment (aditya-nambiar) [inline fennel/test_lib/mock_client.py:80]: Can you add a comment / description to this function

**Review by aditya-nambiar — APPROVED** [review]
Please update the client version + changelog

File: fennel/test_lib/mock_client.py (hunk 5 above)
Comment (nonibansal) [reply fennel/test_lib/mock_client.py:80]: yup added.

**Pull request merged**